rfc/0018-certificate-handling.md

```diff
@@ -0,0 +1,259 @@
+> instances use a certificate that is signed by our CA. Hence we always put our CA
+> inside of the `clientConfig.caBundle`.
+
+This is already done by our current code (Kubewarden 1.6.0).
+
+### Policy Server Certificate Renewed
+
+From time to time we need to renew the certificate issued to a Policy Server instance. The main reason
+to perform this operation is to ensure the Policy Server is not using an expired certificate.
+
+To renew a Policy Server certificate the following actions have to be done:
+
+- Generate new certificate, sign it with our CA
+- Force a rollout of the Policy Server deployment. This will ensure the HTTPS server uses the new certificate
+
+> **Note:** there's no need to touch the webhook configuration objects related to the policies
+> hosted by the Policy Server. The new certificate is still signed by the same CA, hence nothing has
+> to be changed.
+
+### Internal CA changes
+
+The internal CA managed by Kubewarden might be recreated (for example, to avoid its expiration). This is the
+most disruptive event that can happen, especially once some policies are deployed.
+
+Once the new CA is generated, the following actions have to be performed:
+
+- For each `PolicyServer` defined:
+  - Generate a new certificate, sign it with the new CA
+  - Force a rollout of the PolicyServer Deployment
+- Kubewarden Controller:
+  - Generate a new certificate, sign it with the new CA
+  - Restart the controller to ensure its HTTPS endpoint uses the new certificate
+- For each webhook configuration managed by Kubewarden (meaning all the policies deployed, plus the kubewarden-controller):
+  - Update the `clientConfig.caBundle`: ensure it contains the certificate of the new CA
+
+This change is massive and could lead to connection errors for as long as the PolicyServer are using a certificate
+signed by a CA that is not the one mentioned inside of the `clientConfig.caBundle`.
+
+## Controller Bootstrap
+
+At startup, the kubewarden-controller must ensure the root CA has been created. Currently (Kubewarden 1.6.0), the
+controller creates the root CA only when the 1st PolicyServer object is defined.
+We can keep creating a root CA that expires after 10 years.
+
+The code must also ensure the root CA is not about to expire. When this scenario is detected, the controller must renew
+the root CA.
+
+## Reconciliation Loops
+
+This section describes how the kubewarden-controller reconciliation loops should be arranged.
+
+Currently we have these reconciliation loops:
+
+- `PolicyServer`
+- `ClusterAdmissionPolicy`
+- `AdmissionPolicy`
+
+### Policy Server
+
+Reconciliation loops are triggered whenever an event dealing with a watched resource happens.
+In this case it could be the create/update/delete of a `PolicyServer`.
+The reconciliation loop is also triggered every X seconds as a way to cope with possible glitches
+with the event notification system.
```

Comment:
Where is this time based trigger setup? Is this done by kube-builder? I cannot see this  in the code. =(

Reply:
Good question, we're currently using [`controller-runtime`](https://github.com/kubernetes-sigs/controller-runtime) version 0.14.6.

When creating the Manager we give to it some configuration Options, see [here](https://github.com/kubewarden/kubewarden-controller/blob/fc6e1e815f87a80cedb737ca625604f5e80557d6/main.go#L135)

One of the values of the Manager Options is called [`SyncPeriod`](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.6/pkg/manager#Options). Since we don't specify anything, we will have an automatic sync interval of 10 hours. Meaning, even if there's no event affecting one of our watched resources, we will still have a sync every 10 hours.

This configuration parameter has been moved to another location inside of the latest release of controller-runtime. Now this is one of the [configuration values of the cache](https://github.com/kubernetes-sigs/controller-runtime/blob/d5bc8734caccabddac6a1bea250b0b9d771d318d/pkg/cache/cache.go#L160). The behaviour stays the same: every 10 hours you get a full sync.

We can rely on this 10 hours sync or we could be more explicit inside of our reconciliation loops by returning a  `reconcile.Result{RequeueAfter: t}` instead of a `reconcile.Result{}`.

What do you think?

Reply:
I'd probably go with the latter.  As the docstring states, we shouldn't change the SyncPeriod as it forces every object in the cache to be reconciled, and we might need a shorter window than 10 hours. Not a strong opinion, though.

Reply:
I would either leave things untouched, hence rely on the default sync value, or provide an explicit sync timer via the `Result` response

Reply:
> I'd probably go with the latter. 

Sorry I did not explain myself well, I was meaning that I had a  preference for the Result solution.

Reply:
No worries, that was clear to me. I also share the same feeling about how to move forward.

This change (being explicit about the sync time) should be tracked with a dedicated issue, which would be unrelated to this RFC